File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedItem.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.inventory.PlayerInventory;
import cn.nukkit.item.Item;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.recording.player.PlayerRecordNode;
import net.easecation.ghosty.util.PersistenceBinaryStreamHelper;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 17:02.
 * All rights reserved
 */
public class PlayerUpdatedItem implements PlayerUpdated {

    private Item item;

    public static PlayerUpdatedItem of(Item item) {
        return new PlayerUpdatedItem(item);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_ITEM;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        if (ghost != null && ghost.getInventory() != null) {
            PlayerInventory inv = ghost.getInventory();
            inv.setItemInHand(item == null ? Item.get(Item.AIR) : item);
            inv.sendHeldItem(ghost.getViewers().values());
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setItem(item);
        return node;
    }

    public PlayerUpdatedItem(BinaryStream stream, int formatVersion) {
        switch (formatVersion) {
            case 1, 2: {
                this.item = PersistenceBinaryStreamHelper.getItem(stream);
                break;
            }
            case 0: {
                read(stream);
                break;
            }
            default:
                throw new IllegalArgumentException("Unsupported format version: " + formatVersion);
        }
    }

    private PlayerUpdatedItem(Item item) {
        this.item = item;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedItem o)) return false;
        return (item.equals(o.item));
    }

    @Override
    public void write(BinaryStream stream) {
        PersistenceBinaryStreamHelper.putItem(stream, this.item);
    }

    @Override
    public void read(BinaryStream stream) {
        this.item = stream.getSlot();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedItem{" +
            "item=" + item +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedItem.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.item.Item
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.player.PlayerRecordNode
import net.easecation.ghosty.serializer.NukkitItemSerializer
import net.easecation.ghosty.util.PersistenceBinaryStreamHelper

@Serializable
data class PlayerUpdatedItem(
    @Serializable(with = NukkitItemSerializer::class)
    val item: Item
) : PlayerUpdated, UpdateWithState {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_ITEM

    override fun processTo(ghost: PlaybackNPC) {
        ghost.inventory?.let {
            it.itemInHand = item
            it.sendHeldItem(ghost.viewers.values)
        }
    }

    override fun applyTo(node: PlayerRecordNode) = node.apply { this.item = this@PlayerUpdatedItem.item }

    companion object {
        @JvmStatic
        fun of(item: Item) = PlayerUpdatedItem(item)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedItem> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedItem> {
        override fun write(updated: PlayerUpdatedItem, stream: BinaryStream) {
            PersistenceBinaryStreamHelper.putItem(stream, updated.item)
        }

        override fun read(stream: BinaryStream): PlayerUpdatedItem = read(stream, 3)

        override fun read(stream: BinaryStream, version: Int): PlayerUpdatedItem {
            return when (version) {
                0 -> PlayerUpdatedItem(stream.slot)
                1, 2, 3 -> PlayerUpdatedItem(PersistenceBinaryStreamHelper.getItem(stream))
                else -> throw IllegalArgumentException("Unsupported format version: $version")
            }
        }
    }
}
